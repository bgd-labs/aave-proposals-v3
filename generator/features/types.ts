import {Hex} from 'viem';
import {NumberInputValues, PercentInputValues} from '../prompts';
import {BooleanSelectValues} from '../prompts/boolPrompt';

export interface AssetSelector {
  asset: string;
}

export interface TokenImplementations {
  aToken: Hex;
  vToken: Hex;
  sToken: Hex;
}

export interface CapsUpdatePartial {
  supplyCap: NumberInputValues;
  borrowCap: NumberInputValues;
}

export interface CapsUpdate extends CapsUpdatePartial, AssetSelector {}

export interface BorrowUpdatePartial {
  enabledToBorrow: BooleanSelectValues;
  flashloanable: BooleanSelectValues;
  borrowableInIsolation: BooleanSelectValues;
  withSiloedBorrowing: BooleanSelectValues;
  reserveFactor: PercentInputValues;
}

export interface BorrowUpdate extends BorrowUpdatePartial, AssetSelector {}

export interface CollateralUpdatePartial {
  ltv: PercentInputValues;
  liqThreshold: PercentInputValues;
  liqBonus: PercentInputValues;
  debtCeiling: NumberInputValues;
  liqProtocolFee: PercentInputValues;
}

export interface CollateralUpdate extends CollateralUpdatePartial, AssetSelector {}

export interface PriceFeedUpdatePartial {
  priceFeed: Hex;
}

export interface PriceFeedUpdate extends PriceFeedUpdatePartial, AssetSelector {}

export interface AssetEModeUpdatePartial {
  eModeCategory: string;
  collateral: BooleanSelectValues;
  borrowable: BooleanSelectValues;
}

export interface AssetEModeUpdate extends AssetEModeUpdatePartial, AssetSelector {}

export interface EModeCategoryUpdate {
  // library accessor or new id
  eModeCategory: string | number;
  ltv: NumberInputValues;
  liqThreshold: NumberInputValues;
  liqBonus: NumberInputValues;
  priceSource?: Hex | '';
  label: string;
}

export interface RateStrategyParams {
  optimalUtilizationRate: string;
  baseVariableBorrowRate: string;
  variableRateSlope1: string;
  variableRateSlope2: string;
  stableRateSlope1?: string;
  stableRateSlope2?: string;
  baseStableRateOffset?: string;
  stableRateExcessOffset?: string;
  optimalStableToTotalDebtRatio?: string;
}

export interface RateStrategyUpdate extends AssetSelector {
  params: RateStrategyParams;
}

export interface Listing
  extends CollateralUpdatePartial,
    BorrowUpdatePartial,
    CapsUpdatePartial,
    PriceFeedUpdatePartial {
  asset: Hex;
  assetSymbol: string;
  rateStrategyParams: RateStrategyParams;
  eModeCategory: string;
  decimals: number;
  admin?: Hex | '';
}

export interface ListingWithCustomImpl {
  base: Listing;
  implementations: TokenImplementations;
}

export interface TokenStream {
  asset: Hex;
  receiver: Hex;
  duration: string;
  amount: string;
}

export interface FreezeUpdate extends AssetSelector {
  shouldBeFrozen: boolean;
}

export interface EmissionUpdate {
  asset: Hex;
  symbol: string;
  admin: Hex;
}
