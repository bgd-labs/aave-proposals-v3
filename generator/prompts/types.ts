import {PoolIdentifier} from '../types';

export interface GenericPrompt<T extends boolean = boolean> {
  message: string;
  required?: T;
  transform?: (value: string) => string;
  defaultValue?: string;
}

export interface GenericPoolPrompt<T extends boolean = boolean> extends GenericPrompt<T> {
  pool: PoolIdentifier;
}
