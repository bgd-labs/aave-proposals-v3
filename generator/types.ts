import * as addressBook from '@bgd-labs/aave-address-book';
import {
  AssetEModeUpdate,
  BorrowUpdate,
  BorrowUpdatePartial,
  CapsUpdate,
  CollateralUpdate,
  EModeCategoryUpdate,
  Listing,
  ListingWithCustomImpl,
  PriceFeedUpdate,
  RateStrategyUpdate,
  FreezeUpdate,
} from './features/types';
import {FlashBorrower} from './features/flashBorrower';

export const V2_POOLS = [
  'AaveV2Ethereum',
  'AaveV2EthereumAMM',
  'AaveV2Polygon',
  'AaveV2Avalanche',
] as const satisfies readonly (keyof typeof addressBook)[];

export const V3_POOLS = [
  'AaveV3Ethereum',
  'AaveV3EthereumLido',
  'AaveV3EthereumEtherFi',
  'AaveV3Polygon',
  'AaveV3Avalanche',
  'AaveV3Optimism',
  'AaveV3Arbitrum',
  'AaveV3Metis',
  'AaveV3Base',
  'AaveV3Gnosis',
  'AaveV3Scroll',
  'AaveV3BNB',
  'AaveV3ZkSync',
] as const satisfies readonly (keyof typeof addressBook)[];

export const POOLS = [
  ...V2_POOLS,
  ...V3_POOLS,
] as const satisfies readonly (keyof typeof addressBook)[];

export type PoolIdentifier = (typeof POOLS)[number];
export type PoolIdentifierV3 = (typeof V3_POOLS)[number];

export interface Options {
  force?: boolean;
  pools: PoolIdentifier[];
  title: string;
  votingNetwork?: VOTING_NETWORK;
  // automatically generated shortName from title
  shortName: string;
  author: string;
  discussion: string;
  snapshot: string;
  configFile?: string;
  date: string;
}

export type PoolConfigs = Partial<Record<PoolIdentifier, PoolConfig>>;

export type CodeArtifact = {
  code?: {
    constants?: string[];
    fn?: string[];
    execute?: string[];
  };
  test?: {
    fn?: string[];
  };
  aip?: {
    specification: string[];
  };
};

export enum FEATURE {
  ASSET_LISTING = 'ASSET_LISTING',
  ASSET_LISTING_CUSTOM = 'ASSET_LISTING_CUSTOM',
  BORROWS_UPDATE = 'BORROWS_UPDATE',
  CAPS_UPDATE = 'CAPS_UPDATE',
  COLLATERALS_UPDATE = 'COLLATERALS_UPDATE',
  EMODES_ASSETS = 'EMODES_ASSETS',
  EMODES_UPDATES = 'EMODES_UPDATES',
  FLASH_BORROWER = 'FLASH_BORROWER',
  PRICE_FEEDS_UPDATE = 'PRICE_FEEDS_UPDATE',
  RATE_UPDATE_V3 = 'RATE_UPDATE_V3',
  RATE_UPDATE_V2 = 'RATE_UPDATE_V2',
  FREEZE = 'FREEZE',
  OTHERS = 'OTHERS',
}

export enum VOTING_NETWORK {
  POLYGON = 'POLYGON',
  ETHEREUM = 'ETHEREUM',
  AVALANCHE = 'AVALANCHE',
}

export interface FeatureModule<T extends {} = {}> {
  description: string;
  value: FEATURE;
  cli: (args: {options: Options; pool: PoolIdentifier; cache: PoolCache}) => Promise<T>;
  build: (args: {options: Options; pool: PoolIdentifier; cache: PoolCache; cfg: T}) => CodeArtifact;
}

export const ENGINE_FLAGS = {
  KEEP_CURRENT: 'KEEP_CURRENT',
  KEEP_CURRENT_STRING: 'KEEP_CURRENT_STRING',
  KEEP_CURRENT_ADDRESS: 'KEEP_CURRENT_ADDRESS',
  ENABLED: 'ENABLED',
  DISABLED: 'DISABLED',
} as const;

export const AVAILABLE_VERSIONS = {V2: 'V2', V3: 'V3'} as const;

export type ConfigFile = {
  rootOptions: Options;
  poolOptions: Partial<Record<PoolIdentifier, Omit<PoolConfig, 'artifacts'>>>;
};

export type PoolCache = {blockNumber: number};

export interface PoolConfig {
  artifacts: CodeArtifact[];
  configs: {
    [FEATURE.ASSET_LISTING]?: Listing[];
    [FEATURE.ASSET_LISTING_CUSTOM]?: ListingWithCustomImpl[];
    [FEATURE.BORROWS_UPDATE]?: BorrowUpdate[];
    [FEATURE.CAPS_UPDATE]?: CapsUpdate[];
    [FEATURE.COLLATERALS_UPDATE]?: CollateralUpdate[];
    [FEATURE.EMODES_ASSETS]?: AssetEModeUpdate[];
    [FEATURE.EMODES_UPDATES]?: EModeCategoryUpdate[];
    [FEATURE.FLASH_BORROWER]?: FlashBorrower;
    [FEATURE.PRICE_FEEDS_UPDATE]?: PriceFeedUpdate[];
    [FEATURE.RATE_UPDATE_V3]?: RateStrategyUpdate[]; // TODO: type could be improved
    [FEATURE.RATE_UPDATE_V2]?: RateStrategyUpdate[];
    [FEATURE.FREEZE]?: FreezeUpdate[];
    [FEATURE.OTHERS]?: {};
  };
  cache: PoolCache;
}

export type Scripts = {
  defaultScript: string;
  zkSyncScript?: string;
};

export type Files = {
  jsonConfig: string;
  scripts: Scripts;
  aip: string;
  payloads: {pool: PoolIdentifier; payload: string; test: string; contractName: string}[];
};
