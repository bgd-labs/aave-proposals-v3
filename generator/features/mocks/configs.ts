import {Options} from '../../types';
import {CapsUpdate, Listing} from '../types';

export const MOCK_OPTIONS: Options = {
  pools: ['AaveV3Ethereum'],
  title: 'test',
  shortName: 'Test',
  date: '20231023',
  author: 'test',
  discussion: 'test',
  snapshot: 'test',
};

export const assetListingConfig: Listing[] = [
  {
    assetSymbol: 'PSP',
    decimals: 18,
    priceFeed: '0x72AFAECF99C9d9C8215fF44C77B94B99C28741e8',
    ltv: '40_00',
    liqThreshold: '50_00',
    liqBonus: '5_00',
    debtCeiling: '100_000',
    liqProtocolFee: '20_00',
    enabledToBorrow: 'ENABLED',
    flashloanable: 'ENABLED',
    stableRateModeEnabled: 'DISABLED',
    borrowableInIsolation: 'DISABLED',
    withSiloedBorrowing: 'DISABLED',
    reserveFactor: '20_00',
    supplyCap: '10_000',
    borrowCap: '5_000',
    rateStrategyParams: {
      optimalUtilizationRate: '_bpsToRay(80_00)',
      baseVariableBorrowRate: '_bpsToRay(0_00)',
      variableRateSlope1: '_bpsToRay(10_00)',
      variableRateSlope2: '_bpsToRay(100_00)',
      stableRateSlope1: '_bpsToRay(10_00)',
      stableRateSlope2: '_bpsToRay(100_00)',
      baseStableRateOffset: '_bpsToRay(1_00)',
      stableRateExcessOffset: '_bpsToRay()',
      optimalStableToTotalDebtRatio: '_bpsToRay(10_00)',
    },
    eModeCategory: 'AaveV3EthereumEModes.NONE',
    asset: '0xcAfE001067cDEF266AfB7Eb5A286dCFD277f3dE5',
  },
];
