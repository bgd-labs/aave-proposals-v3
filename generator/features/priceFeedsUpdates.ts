import {CodeArtifact, FEATURE, FeatureModule} from '../types';
import {PriceFeedUpdate, PriceFeedUpdatePartial} from './types';
import {addressPrompt, translateJsAddressToSol} from '../prompts/addressPrompt';
import {
  assetsSelectPrompt,
  translateAssetToAssetLibUnderlying,
} from '../prompts/assetsSelectPrompt';

async function fetchPriceFeedUpdate(): Promise<PriceFeedUpdatePartial> {
  return {
    priceFeed: await addressPrompt({
      message: 'New price feed address',
      required: true,
    }),
  };
}

export const priceFeedsUpdates: FeatureModule<PriceFeedUpdate[]> = {
  value: FEATURE.PRICE_FEEDS_UPDATE,
  description: 'PriceFeedsUpdates (replacing priceFeeds)',
  async cli(opt, pool) {
    const response: PriceFeedUpdate[] = [];
    const assets = await assetsSelectPrompt({
      message: 'Select the assets you want to amend',
      pool,
    });
    for (const asset of assets) {
      console.log(`collecting info for ${asset}`);
      response.push({asset, ...(await fetchPriceFeedUpdate())});
    }
    return response;
  },
  build(opt, pool, cfg) {
    const response: CodeArtifact = {
      code: {
        fn: [
          `function priceFeedsUpdates() public pure override returns (IAaveV3ConfigEngine.PriceFeedUpdate[] memory) {
          IAaveV3ConfigEngine.PriceFeedUpdate[] memory priceFeedUpdates = new IAaveV3ConfigEngine.PriceFeedUpdate[](${
            cfg.length
          });

          ${cfg
            .map(
              (cfg, ix) => `priceFeedUpdates[${ix}] = IAaveV3ConfigEngine.PriceFeedUpdate({
               asset: ${translateAssetToAssetLibUnderlying(cfg.asset, pool)},
               priceFeed: ${translateJsAddressToSol(cfg.priceFeed)}
             });`
            )
            .join('\n')}

          return priceFeedUpdates;
        }`,
        ],
      },
    };
    return response;
  },
};
