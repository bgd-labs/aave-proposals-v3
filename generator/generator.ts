import fs from 'fs';
import path from 'path';
import {generateContractName, generateFolderName} from './common';
import {proposalTemplate} from './templates/proposal.template';
import {testTemplate} from './templates/test.template';
import {confirm} from '@inquirer/prompts';
import {ConfigFile, Options, PoolConfigs, PoolIdentifier} from './types';
import prettier from 'prettier';
import {generateScript} from './templates/script.template';
import {generateAIP} from './templates/aip.template';

const prettierSolCfg = await prettier.resolveConfig('foo.sol');
const prettierMDCfg = await prettier.resolveConfig('foo.md');
const prettierTsCfg = await prettier.resolveConfig('foo.ts');

type Files = {
  jsonConfig: string;
  script: string;
  aip: string;
  payloads: {payload: string; test: string; contractName: string}[];
};

/**
 * Generates all the file contents for aip/tests/payloads & script
 * @param options
 * @param poolConfigs
 * @returns
 */
export async function generateFiles(options: Options, poolConfigs: PoolConfigs): Promise<Files> {
  const jsonConfig = prettier.format(
    `import {ConfigFile} from '../../generator/types';
    export const config: ConfigFile = ${JSON.stringify({
      rootOptions: options,
      poolOptions: (Object.keys(poolConfigs) as PoolIdentifier[]).reduce((acc, pool) => {
        acc[pool] = {configs: poolConfigs[pool]!.configs, cache: poolConfigs[pool]!.cache};
        return acc;
      }, {}),
    } as ConfigFile)}`,
    {...prettierTsCfg, filepath: 'foo.ts'}
  );

  function createPayloadAndTest(options: Options, pool: PoolIdentifier) {
    const contractName = generateContractName(options, pool);
    const testCode = testTemplate(options, poolConfigs[pool]!, pool);
    return {
      payload: prettier.format(proposalTemplate(options, poolConfigs[pool]!, pool), {
        ...prettierSolCfg,
        filepath: 'foo.sol',
      }),
      test: prettier.format(testCode, {
        ...prettierSolCfg,
        filepath: 'foo.sol',
      }),
      contractName: contractName,
    };
  }

  console.log('generating script');
  const script = prettier.format(generateScript(options), {
    ...prettierSolCfg,
    filepath: 'foo.sol',
  });
  console.log('generating aip');
  const aip = prettier.format(generateAIP(options, poolConfigs), {
    ...prettierMDCfg,
    filepath: 'foo.md',
  });

  return {
    jsonConfig,
    script,
    aip,
    payloads: options.pools.map((pool) => createPayloadAndTest(options, pool)),
  };
}

async function askBeforeWrite(options: Options, path: string, content: string) {
  if (!options.force && fs.existsSync(path)) {
    const force = await confirm({
      message: `A file already exists at ${path} do you want to overwrite`,
      default: false,
    });
    if (!force) return;
  }
  fs.writeFileSync(path, content);
}

/**
 * Writes the files according to defined folder/file format
 * @param options
 * @param param1
 */
export async function writeFiles(options: Options, {jsonConfig, script, aip, payloads}: Files) {
  const baseName = generateFolderName(options);
  const baseFolder = path.join(process.cwd(), 'src', baseName);
  if (fs.existsSync(baseFolder)) {
    if (!options.force && fs.existsSync(baseFolder)) {
      const force = await confirm({
        message: 'A proposal already exists at that location, do you want to continue?',
        default: false,
      });
      if (!force) return;
    }
  } else {
    fs.mkdirSync(baseFolder, {recursive: true});
  }

  // write config
  await askBeforeWrite(options, path.join(baseFolder, 'config.ts'), jsonConfig);
  // write aip
  await askBeforeWrite(options, path.join(baseFolder, `${options.shortName}.md`), aip);
  // write scripts
  await askBeforeWrite(
    options,
    path.join(baseFolder, `${generateContractName(options)}.s.sol`),
    script
  );

  for (const {payload, test, contractName} of payloads) {
    await askBeforeWrite(options, path.join(baseFolder, `${contractName}.sol`), payload);
    await askBeforeWrite(options, path.join(baseFolder, `${contractName}.t.sol`), test);
  }
}
