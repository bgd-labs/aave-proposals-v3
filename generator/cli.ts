import path from 'path';
import {Command, Option} from 'commander';
import {CHAIN_TO_CHAIN_ID, getDate, getPoolChain, isV2Pool, pascalCase} from './common';
import {input, checkbox, select} from '@inquirer/prompts';
import {
  CodeArtifact,
  ConfigFile,
  FEATURE,
  FeatureModule,
  Options,
  POOLS,
  PoolCache,
  PoolConfigs,
  PoolIdentifier,
  VOTING_NETWORK,
} from './types';
import {flashBorrower} from './features/flashBorrower';
import {capsUpdates} from './features/capsUpdates';
import {rateUpdatesV2, rateUpdatesV3} from './features/rateUpdates';
import {collateralsUpdates} from './features/collateralsUpdates';
import {borrowsUpdates} from './features/borrowsUpdates';
import {eModeUpdates} from './features/eModesUpdates';
import {eModeAssets} from './features/eModesAssets';
import {priceFeedsUpdates} from './features/priceFeedsUpdates';
import {freezeUpdates} from './features/freeze';
import {emissionUpdates} from './features/emission';
import {assetListing, assetListingCustom} from './features/assetListing';
import {generateFiles, writeFiles} from './generator';
import {CHAIN_ID_CLIENT_MAP} from '@bgd-labs/js-utils';
import {getBlockNumber} from 'viem/actions';

const program = new Command();

program
  .name('proposal-generator')
  .description('CLI to generate aave proposals')
  .version('1.0.0')
  .addOption(new Option('-f, --force', 'force creation (might overwrite existing files)'))
  .addOption(new Option('-p, --pools <pools...>').choices(POOLS))
  .addOption(new Option('-t, --title <string>', 'aip title'))
  .addOption(new Option('-a, --author <string>', 'author'))
  .addOption(new Option('-d, --discussion <string>', 'forum link'))
  .addOption(new Option('-s, --snapshot <string>', 'snapshot link'))
  .addOption(
    new Option(
      '-v, --votingNetwork <votingNetwork...>',
      'network where voting should take place for the proposal',
    ).choices(Object.values(VOTING_NETWORK)),
  )
  .addOption(new Option('-c, --configFile <string>', 'path to config file'))
  .allowExcessArguments(false)
  .parse(process.argv);

let options = program.opts<Options>();
let poolConfigs: PoolConfigs = {};

const PLACEHOLDER_MODULE: FeatureModule<{}> = {
  description: 'Something different not supported by configEngine',
  value: FEATURE.OTHERS,
  cli: async ({}) => {
    return {};
  },
  build: ({}) => {
    const response: CodeArtifact = {
      code: {execute: ['// custom code goes here']},
    };
    return response;
  },
};
const FEATURE_MODULES_V2 = [rateUpdatesV2, PLACEHOLDER_MODULE];
const FEATURE_MODULES_V3 = [
  rateUpdatesV3,
  capsUpdates,
  collateralsUpdates,
  borrowsUpdates,
  flashBorrower,
  priceFeedsUpdates,
  eModeUpdates,
  eModeAssets,
  assetListing,
  assetListingCustom,
  freezeUpdates,
  emissionUpdates,
  PLACEHOLDER_MODULE,
];

async function generateDeterministicPoolCache(pool: PoolIdentifier): Promise<PoolCache> {
  const chain = getPoolChain(pool);
  const client = CHAIN_ID_CLIENT_MAP[CHAIN_TO_CHAIN_ID[chain]];
  return {blockNumber: Number(await getBlockNumber(client))};
}

async function fetchPoolOptions(pool: PoolIdentifier) {
  poolConfigs[pool] = {
    configs: {},
    artifacts: [],
    cache: await generateDeterministicPoolCache(pool),
  };

  const v2 = isV2Pool(pool);
  const features = await checkbox({
    message: `What do you want to do on ${pool}?`,
    choices: v2
      ? FEATURE_MODULES_V2.map((m) => ({value: m.value, name: m.description}))
      : FEATURE_MODULES_V3.map((m) => ({value: m.value, name: m.description})),
  });
  for (const feature of features) {
    const module = v2
      ? FEATURE_MODULES_V2.find((m) => m.value === feature)!
      : FEATURE_MODULES_V3.find((m) => m.value === feature)!;
    poolConfigs[pool]!.configs[feature] = await module.cli({
      options,
      pool,
      cache: poolConfigs[pool]!.cache,
    });
    poolConfigs[pool]!.artifacts.push(
      module.build({
        options,
        pool,
        cfg: poolConfigs[pool]!.configs[feature],
        cache: poolConfigs[pool]!.cache,
      }),
    );
  }
}

if (options.configFile) {
  const {config: cfgFile}: {config: ConfigFile} = await import(
    path.join(process.cwd(), options.configFile)
  );
  options = {...options, ...cfgFile.rootOptions};
  poolConfigs = cfgFile.poolOptions as any;
  for (const pool of options.pools) {
    const v2 = isV2Pool(pool);
    if (poolConfigs[pool]) {
      poolConfigs[pool]!.artifacts = [];
      for (const feature of Object.keys(poolConfigs[pool]!.configs)) {
        const module = v2
          ? FEATURE_MODULES_V2.find((m) => m.value === feature)!
          : FEATURE_MODULES_V3.find((m) => m.value === feature)!;
        poolConfigs[pool]!.artifacts.push(
          module.build({
            options,
            pool,
            cfg: poolConfigs[pool]!.configs[feature],
            cache: poolConfigs[pool]!.cache,
          }),
        );
      }
    } else {
      await fetchPoolOptions(pool);
    }
  }
} else {
  options.pools = await checkbox({
    message: 'Chains this proposal targets',
    choices: POOLS.map((v) => ({name: v, value: v})),
    required: true,
    // validate(input) {
    //   // currently ignored due to a bug
    //   if (input.length == 0) return 'You must target at least one chain in your proposal!';
    //   return true;
    // },
  });

  if (!options.title) {
    options.title = await input({
      message:
        'Short title of your proposal that will be used as contract name(please refrain from including author or date)',
      validate(input) {
        if (input.length == 0) return "Your title can't be empty";
        // this is no exact math
        // fully qualified identifiers are not allowed to be longer then 300 chars on etherscan api
        // the path is roughly src(3)/date(8)_title/title_date(8):title_date(8), so 3 + 3*8 + 3 title.length
        // so 80 sounds like a reasonable upper bound to stay below 300 character limit
        if (input.trim().length > 80) return 'Your title is to long';
        return true;
      },
    });
  }
  options.shortName = pascalCase(options.title);
  options.date = getDate();

  if (!options.author) {
    options.author = await input({
      message: 'Author of your proposal',
      validate(input) {
        if (input.length == 0) return "Your author can't be empty";
        return true;
      },
    });
  }

  if (!options.discussion) {
    options.discussion = await input({
      message: 'Link to forum discussion',
    });
  }

  if (!options.snapshot) {
    options.snapshot = await input({
      message: 'Link to snapshot',
    });
  }

  if (!options.votingNetwork) {
    options.votingNetwork = await select({
      message: 'Select network where voting should takes place for the proposal',
      choices: Object.values(VOTING_NETWORK).map((v) => ({
        name: v != VOTING_NETWORK.POLYGON ? v : VOTING_NETWORK.POLYGON + ' (DEFAULT)',
        value: v,
      })),
      default: VOTING_NETWORK.POLYGON,
    });
  }

  for (const pool of options.pools) {
    await fetchPoolOptions(pool);
  }
}

try {
  const files = await generateFiles(options, poolConfigs);
  await writeFiles(options, files);
} catch (e) {
  console.log(JSON.stringify({options, poolConfigs}, null, 2));
  throw e;
}
