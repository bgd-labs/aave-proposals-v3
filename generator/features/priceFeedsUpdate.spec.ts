// sum.test.js
import {expect, describe, it} from 'vitest';
import {MOCK_OPTIONS, priceFeedsUpdateConfig} from './mocks/configs';
import {generateFiles} from '../generator';
import {FEATURE, PoolConfigs} from '../types';
import {priceFeedsUpdates} from './priceFeedsUpdates';

describe('feature: priceFeedsUpdates', () => {
  it('should return reasonable code', () => {
    const output = priceFeedsUpdates.build(MOCK_OPTIONS, 'AaveV3Ethereum', priceFeedsUpdateConfig);
    expect(output).toMatchSnapshot();
  });

  it('should properly generate files', async () => {
    const poolConfigs: PoolConfigs = {
      [MOCK_OPTIONS.pools[0]]: {
        pool: MOCK_OPTIONS.pools[0],
        features: [FEATURE.PRICE_FEEDS_UPDATE],
        artifacts: [
          priceFeedsUpdates.build(MOCK_OPTIONS, 'AaveV3Ethereum', priceFeedsUpdateConfig),
        ],
        configs: {[FEATURE.PRICE_FEEDS_UPDATE]: priceFeedsUpdateConfig},
      },
    };
    const files = await generateFiles(MOCK_OPTIONS, poolConfigs);
    expect(files).toMatchSnapshot();
  });
});
