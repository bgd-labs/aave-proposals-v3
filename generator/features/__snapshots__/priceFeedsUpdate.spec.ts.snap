// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`feature: priceFeedsUpdates > should properly generate files 1`] = `
{
  "aip": "---
title: \\"test\\"
author: \\"test\\"
discussions: \\"test\\"
---

## Simple Summary

## Motivation

## Specification

## References

- Implementation: [AaveV3Ethereum](https://github.com/bgd-labs/aave-proposals-v3/blob/main/src/20231023_AaveV3Ethereum_Test/AaveV3Ethereum_Test_20231023.sol)
- Tests: [AaveV3Ethereum](https://github.com/bgd-labs/aave-proposals-v3/blob/main/src/20231023_AaveV3Ethereum_Test/AaveV3Ethereum_Test_20231023.t.sol)
- [Snapshot](test)
- [Discussion](test)

## Copyright

Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
",
  "jsonConfig": "import {ConfigFile} from '../../generator/types';
export const config: ConfigFile = {
  rootOptions: {
    pools: ['AaveV3Ethereum'],
    title: 'test',
    shortName: 'Test',
    date: '20231023',
    author: 'test',
    discussion: 'test',
    snapshot: 'test',
  },
  poolOptions: {
    AaveV3Ethereum: {
      configs: {
        PRICE_FEEDS_UPDATE: [
          {asset: 'DAI', priceFeed: '0xae7ab96520de3a18e5e111b5eaab095312d7fe84'},
        ],
      },
      cache: {blockNumber: 42},
    },
  },
};
",
  "payloads": [
    {
      "contractName": "AaveV3Ethereum_Test_20231023",
      "payload": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {AaveV3EthereumAssets} from 'aave-address-book/AaveV3Ethereum.sol';
import {AaveV3PayloadEthereum} from 'aave-helpers/v3-config-engine/AaveV3PayloadEthereum.sol';
import {IAaveV3ConfigEngine} from 'aave-helpers/v3-config-engine/IAaveV3ConfigEngine.sol';

/**
 * @title test
 * @author test
 * - Snapshot: test
 * - Discussion: test
 */
contract AaveV3Ethereum_Test_20231023 is AaveV3PayloadEthereum {
  function priceFeedsUpdates()
    public
    pure
    override
    returns (IAaveV3ConfigEngine.PriceFeedUpdate[] memory)
  {
    IAaveV3ConfigEngine.PriceFeedUpdate[]
      memory priceFeedUpdates = new IAaveV3ConfigEngine.PriceFeedUpdate[](1);

    priceFeedUpdates[0] = IAaveV3ConfigEngine.PriceFeedUpdate({
      asset: AaveV3EthereumAssets.DAI_UNDERLYING,
      priceFeed: 0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84
    });

    return priceFeedUpdates;
  }
}
",
      "test": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {AaveV3Ethereum} from 'aave-address-book/AaveV3Ethereum.sol';

import 'forge-std/Test.sol';
import {ProtocolV3TestBase, ReserveConfig} from 'aave-helpers/ProtocolV3TestBase.sol';
import {AaveV3Ethereum_Test_20231023} from './AaveV3Ethereum_Test_20231023.sol';

/**
 * @dev Test for AaveV3Ethereum_Test_20231023
 * command: make test-contract filter=AaveV3Ethereum_Test_20231023
 */
contract AaveV3Ethereum_Test_20231023_Test is ProtocolV3TestBase {
  AaveV3Ethereum_Test_20231023 internal proposal;

  function setUp() public {
    vm.createSelectFork(vm.rpcUrl('mainnet'), 42);
    proposal = new AaveV3Ethereum_Test_20231023();
  }

  /**
   * @dev executes the generic test suite including e2e and config snapshots
   */
  function test_defaultProposalExecution() public {
    defaultTest('AaveV3Ethereum_Test_20231023', AaveV3Ethereum.POOL, address(proposal));
  }
}
",
    },
  ],
  "script": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {GovV3Helpers, IPayloadsControllerCore, PayloadsControllerUtils} from 'aave-helpers/GovV3Helpers.sol';
import {EthereumScript} from 'aave-helpers/ScriptUtils.sol';
import {AaveV3Ethereum_Test_20231023} from './AaveV3Ethereum_Test_20231023.sol';

/**
 * @dev Deploy Ethereum
 * deploy-command: make deploy-ledger contract=src/20231023_AaveV3Ethereum_Test/Test_20231023.s.sol:DeployEthereum chain=mainnet
 * verify-command: npx catapulta-verify -b broadcast/Test_20231023.s.sol/1/run-latest.json
 */
contract DeployEthereum is EthereumScript {
  function run() external broadcast {
    // deploy payloads
    AaveV3Ethereum_Test_20231023 payload0 = GovV3Helpers.deployDeterministic(
      type(AaveV3Ethereum_Test_20231023).creationCode
    );

    // compose action
    IPayloadsControllerCore.ExecutionAction[]
      memory actions = new IPayloadsControllerCore.ExecutionAction[](1);
    actions[0] = GovV3Helpers.buildAction(address(payload0));

    // register action at payloadsController
    GovV3Helpers.createPayload(actions);
  }
}

/**
 * @dev Create Proposal
 * command: make deploy-ledger contract=src/20231023_AaveV3Ethereum_Test/Test_20231023.s.sol:CreateProposal chain=mainnet
 */
contract CreateProposal is EthereumScript {
  function run() external {
    // create payloads
    PayloadsControllerUtils.Payload[] memory payloads = new PayloadsControllerUtils.Payload[](1);

    // compose actions for validation
    IPayloadsControllerCore.ExecutionAction[]
      memory actionsEthereum = new IPayloadsControllerCore.ExecutionAction[](1);
    actionsEthereum[0] = GovV3Helpers.buildAction(type(AaveV3Ethereum_Test_20231023).creationCode);
    payloads[0] = GovV3Helpers.buildMainnetPayload(vm, actionsEthereum);

    // create proposal
    vm.startBroadcast();
    GovV3Helpers.createProposal2_5(
      vm,
      payloads,
      GovV3Helpers.ipfsHashFile(vm, 'src/20231023_AaveV3Ethereum_Test/Test.md')
    );
  }
}
",
}
`;

exports[`feature: priceFeedsUpdates > should return reasonable code 1`] = `
{
  "code": {
    "fn": [
      "function priceFeedsUpdates() public pure override returns (IAaveV3ConfigEngine.PriceFeedUpdate[] memory) {
          IAaveV3ConfigEngine.PriceFeedUpdate[] memory priceFeedUpdates = new IAaveV3ConfigEngine.PriceFeedUpdate[](1);

          priceFeedUpdates[0] = IAaveV3ConfigEngine.PriceFeedUpdate({
               asset: AaveV3EthereumAssets.DAI_UNDERLYING,
               priceFeed: 0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84
             });

          return priceFeedUpdates;
        }",
    ],
  },
}
`;
